[#3] 기본 로그인 및 소셜 로그인 구현

ODooHo opened the pull request:
<!-- 제목은 `[#이슈번호] 제목`으로 작성해주세요. (ex) [#1] 프로젝트 초기 설정 작업 -->

### 🧑‍💻 작업 내용
- 회원가입 여정에 가게 정보와 사용자 정보를 동시에 다룸으로 인해서 패키징 분리 및 로직 분리를 수행함.
    - 회원가입을 담당하는 `SignUpService`에서 `AuthService`를 호출하여 사용자를 생성
    - `StoreService`를 호출하여 가게 정보를 생성
    - 이들은 한 트랜잭션에 묶인다.
- 이 과정중 생기는 도메인 별 `ErrorCode`를 정의함
    - `DomainException` 클래스에서 예외를 전담한다.
    - 도메인 별 `ErrorCode`를 추상화하여 예외 메시지 및 처리를 달리 진행한다.

### ✨ 리뷰 포인트
`spotless` 설정 변경한 것 때메 커밋하는 클래스 양이 늘었어요! feat 브랜치만 봐주시면 될 것 같네요

### 🎯 관련 이슈

closed #3

ODooHo committed:
build(auth-3): 소셜 로그인 구현을 위한 의존성 및 프로퍼티 추가

- OAuth-client 의존성 활성화
- jjwt, Jasypt 의존성 추가
- 로컬 임시 db(로그인 여정 테스트를 위함) 프로퍼티 추가

ODooHo committed:
feat(auth-3): end-to-end 테스트를 위한 테스트 컨트롤러 추가

- 임의로 만든 프론트엔드 화면 -> 백엔드 인증 후 멤버 생성 간의 여정을 테스트하기 위해 만든 컨트롤러 클래스(헬스 체크에 이어서)

ODooHo committed:
feat(auth-3): config 시점 프로퍼티 가져오기 위한 어노테이션 추가

- 어플리케이션 시작 시점에 프로퍼티를 일괄적으로 가져오기 위한 어노테이션 추가

ODooHo committed:
feat(auth-3): 소셜 로그인 초기 구현

- 고용주 사용자를 기반으로 소셜 로그인 인증 구현
    - 알바 사용자가 추가되었을 때 변경이 어떻게 전파될지 고려해야함
- 추후 일반 로그인과 병합하기 위해 확장성을 고려하여 설계 및 개발을 진행함.
- 추후 소셜 로그인 플랫폼 추가를 위해 확장성을 고려하여 설계 및 개발을 진행함.
- oAuth 서버와의 통신은 `HttpInterface`를 활용했고 이를 위해 web-flux 의존성이 추가됨

this closes #3

ODooHo committed:
feat(auth-3): Config 클래스 작성

- Spring-security를 위한 config 클래스 작성
- Jasypt 프로퍼티 암호화를 위한 config 클래스 작성

ODooHo committed:
chore(auth-3): merge from dev

- Gihub Actions 관련 스크립트 머지를 포함한 커밋

ODooHo committed:
build(auth-3): spotless 버전 변경 및 설정 변경

- Java Format에 맞게 정렬하는 명령어 추가

ODooHo committed:
chore(auth-3): spotless 적용

ODooHo committed:
chore(auth-3): spotless 적용

ODooHo committed:
build(auth-3): 카카오 OAuth 키 변경 및spotless 설정 변경

- 카카오 OAuth 게정 중 이메일 확인이 가능한 계정으로 변경(사용자 이름은 여전히 가져오지 못함)
biz계정 활성화 관련 논의 필요

ODooHo committed:
feat(auth-3): 기본 회원가입 및 로그인 여정 개발

- 회원가입 여정에 가게 정보와 사용자 정보를 동시에 다룸으로 인해서 패키징 분리 및 로직 분리를 수행함.
    - 회원가입을 담당하는 `SignUpService`에서 `AuthService`를 호출하여 사용자를 생성
    - `StoreService`를 호출하여 가게 정보를 생성
    - 이들은 한 트랜잭션에 묶인다.
- 이 과정중 생기는 도메인 별 `ErrorCode`를 정의함
    - `DomainException` 클래스에서 예외를 전담한다.
    - 도메인 별 `ErrorCode`를 추상화하여 예외 메시지 및 처리를 달리 진행한다.

ODooHo committed:
feat(auth-3): 기본 회원가입 및 로그인 여정 개발

- 이름이 바뀐 프로덕트들을 푸시하는 커밋

ODooHo committed:
chore(auth-3): contextLoads Test disable

codingbaraGo commented:
패키지 구조가 우리 컨벤션 잡은거랑 좀 다른거같네요
특정 서비스 도메인 아래에 domain이란 패키지는 없었던거같은데 도메인 밑에 entity있고 dto있고...
헷갈려서 보기 너무 힘드네요
컨벤션 맞춰서 구조 수정해주세요

codingbaraGo commented:
global도 패키지 구조가 전혀 이해가 안됩니다

codingbaraGo commented:
+ 로그인 후 프론트에서 리퀘스트 받을때 jwt 토큰 체크해서 권한 주는 부분이 있나요?
못찾겠습니다.

codingbaraGo commented:
또 하나 더 건의사항을 말씀드리자면 로그인 ~ 사업장 등록이 기획에서는 하나의 시퀸스라 쳐도 백엔드 서비스 도메인상으론 확실하게 구분이 되는 별개의 도메인이니 이슈랑 브랜치, PR을 분리하는게 맞는 것같습니다.
보기가 너무 힘드네요.
이번은 일단 코드를 작성했으니 이대로 가고 다음부터는 이슈랑 PR은 도메인별로 분리 부탁드립니다.

ODooHo commented:
> * 로그인 후 프론트에서 리퀘스트 받을때 jwt 토큰 체크해서 권한 주는 부분이 있나요?
>   못찾겠습니다.

Authentication 객체를 만들어두고 사용해서 컨트롤러 단에서 `@AuthenticationPrincipal` 어노테이션 활용해서 사용자 정보 가져옵니다. `JwtAuthenticationFilter`에 해당 내용 있으니 참고해주세요

ODooHo commented:
> 패키지 구조가 우리 컨벤션 잡은거랑 좀 다른거같네요 특정 서비스 도메인 아래에 domain이란 패키지는 없었던거같은데 도메인 밑에 entity있고 dto있고... 헷갈려서 보기 너무 힘드네요 컨벤션 맞춰서 구조 수정해주세요

음 일단 컨벤션이랑 구조가 달라진 점을 사전에 말씀드리지 않은 부분은 죄송합니다. domain이란 패키지를 추가한 이유는 소셜 로그인에서 사용되는 `client`라는 객체들이 추가되면서 이들을 단순히 `client`라는 이름으로 패키징 하자니 실제로 client라는 패키징이 의미하는 바랑 좀 다른 것 같더라구요. 복합체 개념을 사용하면서 이들 또한 사용자의 정보를 담고있고 이를 위해 사용되는 부분이라서 `domain` 이라는 상위 패키지를 둬서 여기서 말하는 client는 도메인과 관련된 부분이라는 점을 좀 명시하고 싶었습니다. 그렇게 되면서 자연스럽게 `entity`와 `dto`가 `domain`이라는 좀 상위 개념 아래로 편입이 되어야하겠다는 생각이 들어서 이렇게 패키징을 진행하게 되었구요.

저 부분이 그냥 분리되어도 된다고 생각하신다면 그 이유도 함께 말씀해주시면 감사하겠습니다.

ODooHo commented:
> 또 하나 더 건의사항을 말씀드리자면 로그인 ~ 사업장 등록이 기획에서는 하나의 시퀸스라 쳐도 백엔드 서비스 도메인상으론 확실하게 구분이 되는 별개의 도메인이니 이슈랑 브랜치, PR을 분리하는게 맞는 것같습니다. 보기가 너무 힘드네요. 이번은 일단 코드를 작성했으니 이대로 가고 다음부터는 이슈랑 PR은 도메인별로 분리 부탁드립니다.

이건 이슈 내에 세부 이슈를 만들면서 작업한 부분이라 이슈 활용하는게 익숙치 않아서 발생한 것 같습니다. 작업이 많이 겹치는 부분이 있고 기존에 사용자 회원가입과 사업장 등록을 분리할 것을 염두하고 작업을 했던 터라서 다음부턴 도메인 별로 브랜치 작업할 예정입니다.
중간에 CICD 구현이 별도 공지 없이 dev랑 병합된 부분이 있어서 차라리 그 때 분리를 한 번 진행했다면 좋았을 것 같네요

ODooHo commented:
> global도 패키지 구조가 전혀 이해가 안됩니다

global쪽 oauth 패키징 부분이, 외부와 통신이 들어가는 부분이 있고 거기서 받아오는 dto, Provider 복합체 등이 있어서 저도 이부분은 패키징을 어떤식으로 해야할지 고민이 많았습니다. 이 부분이 부족했던 부분은 맞는 것 같아서 어떤 식으로 하면 좋을지 얘기 같이 나눠보는게 좋겠네요.

나머지 부분은 사실 모호할 부분이 없다고 생각합니다.

ODooHo commented:
이번 PR에 커밋되는 코드들이 많아서 전반적으로 확인에 어려움을 드렸던 부분은 제가 미숙한 부분이라고 생각이 되네요. 
하지만 의문을 가지시는 부분들 대부분이 PR 드리기 전에 중간 점검 차 브랜치를 퍼블리싱 했던 부분에서 포함이 되었던 부분이에요. 퍼블리싱을 한 목적이 중간에 코드 한 번 땡겨보시고, 의문이시거나 그런 부분들을 한 번 짚고 갔으면 하는 생각에 진행했던 것이었는데, 그 때 확인이 좀 더 되었다면 복잡함은 좀 줄었을거라고 생각이 되네요.

의견 더 주실 부분 있다면 얘기 나눠보고 반영할 부분 반영하도록 하겠습니다.

codingbaraGo commented:
> > 패키지 구조가 우리 컨벤션 잡은거랑 좀 다른거같네요 특정 서비스 도메인 아래에 domain이란 패키지는 없었던거같은데 도메인 밑에 entity있고 dto있고... 헷갈려서 보기 너무 힘드네요 컨벤션 맞춰서 구조 수정해주세요
> 
> 음 일단 컨벤션이랑 구조가 달라진 점을 사전에 말씀드리지 않은 부분은 죄송합니다. domain이란 패키지를 추가한 이유는 소셜 로그인에서 사용되는 `client`라는 객체들이 추가되면서 이들을 단순히 `client`라는 이름으로 패키징 하자니 실제로 client라는 패키징이 의미하는 바랑 좀 다른 것 같더라구요. 복합체 개념을 사용하면서 이들 또한 사용자의 정보를 담고있고 이를 위해 사용되는 부분이라서 `domain` 이라는 상위 패키지를 둬서 여기서 말하는 client는 도메인과 관련된 부분이라는 점을 좀 명시하고 싶었습니다. 그렇게 되면서 자연스럽게 `entity`와 `dto`가 `domain`이라는 좀 상위 개념 아래로 편입이 되어야하겠다는 생각이 들어서 이렇게 패키징을 진행하게 되었구요.
> 
> 저 부분이 그냥 분리되어도 된다고 생각하신다면 그 이유도 함께 말씀해주시면 감사하겠습니다.


제가 이해가 힘들었던 포인트를 몇개 말씀드리겠습니다.

**1. EmployerUser Entity와 Repository가 Auth 도메인 아래에 존재**
일단 EmployerUser의 경우에는 Auth 아래에 존재하는게 아니라 별도의 도메인에 존재해야하지 않나 하는 생각입니다. 유저 도메인의 경우에는 로그인 인증과 관련된 기능 외의 기능이 굉장히 많기 때문에 로그인 관련 기능이 유저 도메인의 하위에 존재하거나 로그인 도메인과 유저 도메인이 같은 레벨에서 별도로 존재할 수 는 있지만 로그인 도메인 하위에 유저 도메인이 존재하는건 조금 부자연스럽지 않을까 해요.

**2. EmployerUser의 Entity와 Repository가 Auth 도메인 아래에 존재하지만 둘이 레벨이 다름. Entity의 경우 Auth - Domain - Entity에 위치해 있지만 Repository의 경우 Auth - Repository에 위치.** 
같은 도메인의 Entity와 Repository는 응당 같은 레벨에 존재해야한다고 생각합니다. 지금과 같이 같은 도메인의 요소들이 분리되게되면 보기도 힘들고 향후 다른 기능을 추가하거나 관련 코드를 수정할때 이런저런 문제가 생길 가능성이 있다 생각합니다.

**3. Auth 아래의 Domain이라는 패키지의 이름이 너무 모호. 이 패키지가 어떤 패키지인지 굉장히 헷갈림. 특히 Domain 아래에 Oauth 관련 클래스와 아마도 커스텀 로그인 리퀘스트를 위한 클래스로 추정되는 LoginReq, LoginRes까지 같이 있고 EmployerUser Entity까지 같이 들어있어서 패키지 구조가 굉장히 뒤죽박죽이라는 느낌이 듬.**
두호님이 짜신 코드를 보는 제 입장에서는 컨벤션을 벗어나서 존재하는 패키지의 이름이 `Domain`과 같은 포괄적이고 부정확한 이름이라면 이 패키지의 용도나 의미가 헷갈리게 됩니다. 컨벤션을 벗어난 구조를 만들때에는 네이밍과 주석을 사용해서 해당 코드들이 어떤 기능을 담당하고 왜 존재하는지를 알기 쉽게 해주는 것이 협업을 하는데에 있어 중요한 요소라고 생각합니다.
또, client라는 패키지의 네이밍이 애매하다고 느끼셨다면 OAuthClient는 어떠신가요? 
사실 저는 유저 도메인과 로그인 관련 도메인을 분리하고(둘이 같은 레벨이던 유저 아래에 로그인이 있던) 소셜 로그인과 커스텀 로그인 관련 코드도 분리를 하는편이 좋지 않을까 싶긴 합니다만 로그인은 두호님이 전담하신 부분이니 두호님이 편하신대로 하시면 될 것 같습니다.
다만, 유저 도메인의 경우는 향후 서비스 전반에서 두루두루 사용될 듯 하니 이 부분은 확실하게 분리를 해주셨으면 좋겠습니다.ㅣ

ODooHo committed:
build(auth-3): 카카오 biz 계정으로 변경함에 따른 프로퍼티 수정

codingbaraGo commented:
> > 또 하나 더 건의사항을 말씀드리자면 로그인 ~ 사업장 등록이 기획에서는 하나의 시퀸스라 쳐도 백엔드 서비스 도메인상으론 확실하게 구분이 되는 별개의 도메인이니 이슈랑 브랜치, PR을 분리하는게 맞는 것같습니다. 보기가 너무 힘드네요. 이번은 일단 코드를 작성했으니 이대로 가고 다음부터는 이슈랑 PR은 도메인별로 분리 부탁드립니다.
> 
> 이건 이슈 내에 세부 이슈를 만들면서 작업한 부분이라 이슈 활용하는게 익숙치 않아서 발생한 것 같습니다. 작업이 많이 겹치는 부분이 있고 기존에 사용자 회원가입과 사업장 등록을 분리할 것을 염두하고 작업을 했던 터라서 다음부턴 도메인 별로 브랜치 작업할 예정입니다. 중간에 CICD 구현이 별도 공지 없이 dev랑 병합된 부분이 있어서 차라리 그 때 분리를 한 번 진행했다면 좋았을 것 같네요


CI/CD같은 경우에는 서비스 코드와는 별개의 깃허브 액션 코드를 짜는거라 따로 코드리뷰 과정을 안거쳤네요. 별도의 인프라 작업이라고 생각해서요. 또, 깃허브 홈페이지에서 깃허브 액션 workflow를 짜면 그냥 바로 dev나 main에 푸쉬되어버립니다.
+ CI/CD다 보니 브랜치에 적용을 한 뒤에 돌려보면서 잘 돌아가는지 체크하고 문제가 있으면 코드를 수정한 뒤 다시 브랜치에 푸쉬해서 수정한 코드를 적용한 뒤 체크하고 하는 과정이 필요하여 일일히 PR을 올린 뒤 기다리기에는 시간적으로 문제가 있지 않을까 했습니다.
사실 CI/CD 구축하면서 생긴 커밋이 10개는 더있었는데 너무 더러워서 커밋내역을 다 지우고 하나로 병합해서 지금 보기엔 커밋이 몇개 없어보인다는 비밀도 있습니다.

이쪽으로 공부를 좀 더 해서 다음에 CI/CD 관련 수정을 할때에는 별개의 CI/CD 테스트 브랜치를 파서 테스트를 한 뒤 한번에 완성된 PR을 올릴 수 있도록 해보겠습니다.

ODooHo committed:
chore(auth-3): api, global 패키징 수정

- api
    - domain 패키지 삭제 및 외부(소셜 로그인)api 관련 클래스 `infra` 패키지로 이동
-global
    - 레이어 명확하게 표시

ODooHo committed:
chore(auth-3): spotless Apply

codingbaraGo commented:
> > global도 패키지 구조가 전혀 이해가 안됩니다
> 
> global쪽 oauth 패키징 부분이, 외부와 통신이 들어가는 부분이 있고 거기서 받아오는 dto, Provider 복합체 등이 있어서 저도 이부분은 패키징을 어떤식으로 해야할지 고민이 많았습니다. 이 부분이 부족했던 부분은 맞는 것 같아서 어떤 식으로 하면 좋을지 얘기 같이 나눠보는게 좋겠네요.
> 
> 나머지 부분은 사실 모호할 부분이 없다고 생각합니다.


OAuth쪽 부분 구조에 관해서 같이 이야기 해보는게 좋을것 같긴 하네요.
그 외에 부분은 깔끔한 것같습니다.

일단 한가지 요청사항은 특정 도메인 하위에 패키지를 만들 때 컨벤션 이외의 패키지에는 domain과 같이 unspecipic한 네이밍을 쓰지 않았으면 합니다. 너무 헷갈려요. authcode의 경우에도 어떤 패키지인지 모르겠습니다.

global쪽의 OAuth 관련 패키지의 경우 하나로 묶지말고 컨벤션의 패키지에 맞게 분리하는건 어떤가요?
예를 들어 KakaoOAuthconfig 클래스는 global - config - oauth - kakao 아래에 위치하는 식입니다.
global - oauth - authcode - domain - kakao - dto 아래 클래스들의 경우 특정 서비스 로직에서 사용하는 듯 하니(소셜로그인이라는 서비스 로직) 글로벌 하위보다는 domain 아래의 소셜로그인 관련 패키지쪽에 위치하는게 맞지 않을까 합니다.

ODooHo committed:
chore(auth-3): auth - user 패키징 분리 및 oauth 관련 패키징 변경

- EmployerUser - EmployeeUser 두가지 사용자 도메인이 존재함에 따른 패키징 변경
    - auth: 로그인, 인증 관련 클래스
    - employeruser: 사장님 사용자 관련 클래스
    - employeeuser: 알바생 사용자 관련 클래스
- global에 속해있던 oauth 패키지 이동
    - 외부 api를 활용함에 따라 global 패키지에 두었지만, 목적이 '인증'에 맞춰져 있음에 따라서 패키징 변경 수행

ODooHo committed:
chore(auth-3): auth - user 패키징 분리 및 oauth 관련 패키징 변경

- EmployerUser - EmployeeUser 두가지 사용자 도메인이 존재함에 따른 패키징 변경
    - auth: 로그인, 인증 관련 클래스
    - employeruser: 사장님 사용자 관련 클래스
    - employeeuser: 알바생 사용자 관련 클래스
- global에 속해있던 oauth 패키지 이동
    - 외부 api를 활용함에 따라 global 패키지에 두었지만, 목적이 '인증'에 맞춰져 있음에 따라서 패키징 변경 수행

ODooHo committed:
chore(auth-3): spotless Apply

codingbaraGo commented:
굿입니다
